### PR TITLE
Allow to use a common Sphinx version for Python 2 and 3.

### PR DESCRIPTION
**Pull request by icemac**

### CODE REVIEW

**Comment by icemac** [comment]
Locally the pure Python versions (Python 2 only) fail to build in `tox` when building `BTrees` because a too new version of `setup tools` is used during the build process of BTrees.

**Comment by jamadden** [comment]
The travis macOS 2.7 failure is due to https://github.com/pypa/virtualenv/issues/1555

The workaround is to alter .travis.yml to use `python -m …` like https://github.com/zopefoundation/zope.interface/pull/174/files#diff-354f30a63fb0907d4ad57269548329e3

File: setup.py [hunk 1]
```diff
@@ -197,7 +197,7 @@ def __str__(self):
           'test': TESTS_REQUIRE,
           'docs': [
               'Sphinx < 2; python_version < "3"',
-              'Sphinx >= 2; python_version >= "3"',
+              'Sphinx; python_version >= "3"',
```
Comment (jamadden) [inline setup.py]: I admit I haven't dug into this yet, so I don't understand the reason we need this only on Python 3. Can't Sphinx 1 still be used on Python 2?

File: tox.ini [hunk 2]
```diff
@@ -9,8 +9,11 @@ envlist =
 commands =
     zope-testrunner --test-path=src []
     sphinx-build -b doctest -d {envdir}/.cache/doctrees docs {envdir}/.cache/doctest
+extras =
+    test
+    docs
 deps =
-    .[test,docs]
+    py27,py27-watch,pypy: sphinxcontrib-websupport < 1.2
```
Comment (jamadden) [inline tox.ini]: It's not clear to me why this is here. I don't see us using that extension anywhere?

File: setup.py (hunk 1 above)
Comment (icemac) [reply setup.py]: The line before requires `Sphinx 1.x` for Python 2. Python 3 no longer has a version requirement to be able to pin a Sphinx 1.x version in zopetoolkit.

File: tox.ini (hunk 2 above)
Comment (icemac) [reply tox.ini]: It is a direct dependency of Sphinx (see https://github.com/sphinx-doc/sphinx/blob/591bdd74c0739269a9b2651aec2fa2554a3d8d23/setup.py#L30 for Sphinx 1.8.5) version 1.2 is Python 3 only.

File: setup.py (hunk 1 above)
Comment (jamadden) [reply setup.py]: I apologize, I was on the go between meetings and completely misread the diff.

**Comment by jamadden** [comment]
Removing the version pin entirely works for me in tox:
```patch
@@ -196,8 +196,7 @@ setup(name='zope.security',
           ],
           'test': TESTS_REQUIRE,
           'docs': [
-              'Sphinx < 2; python_version < "3"',
-              'Sphinx >= 2; python_version >= "3"',
+              'Sphinx',
               'repoze.sphinx.autointerface',
           ],
       },
```

```console
$ tox -r -e py27
GLOB sdist-make: …
py27 run-test: commands[0] | zope-testrunner --test-path=src
Running zope.testrunner.layer.UnitTests tests:
  Set up zope.testrunner.layer.UnitTests in 0.000 seconds.
  Ran 753 tests with 0 failures, 0 errors and 0 skipped in 0.266 seconds.
…
Running Sphinx v1.8.5
…
Doctest summary
===============
  514 tests
    0 failures in tests
    0 failures in setup code
    0 failures in cleanup code
build succeeded.
```

Both Sphinx and sphinxcontrib-websupport has the appropriate `python_requires` metadata so that Python 2 will automatically install Sphinx < 2 and sphinxcontrib-websupport < 1.2. I don't see that version pins should be needed at all.

```console
$ .tox/py27/bin/pip freeze | grep -i sphinx
repoze.sphinx.autointerface==0.8
Sphinx==1.8.5
sphinxcontrib-websupport==1.1.2
```

I guess you had a different experience?

**Comment by icemac** [comment]
@jamadden That's strange:

With

```diff
--- a/setup.py
+++ b/setup.py
@@ -196,8 +196,7 @@ setup(name='zope.security',
           ],
           'test': TESTS_REQUIRE,
           'docs': [
-              'Sphinx < 2; python_version < "3"',
-              'Sphinx; python_version >= "3"',
+              'Sphinx',
               'repoze.sphinx.autointerface',
           ],
       },
diff --git a/tox.ini b/tox.ini
index fb50343..51c5d44 100644
--- a/tox.ini
+++ b/tox.ini
@@ -13,7 +13,6 @@ extras =
     test
     docs
 deps =
-    py27,py27-watch,pypy: sphinxcontrib-websupport < 1.2

 [testenv:py27-watch]
 basepython =
```

I get the following error when running the Python 2.7 emvironment:

```
$ tox -repy27
GLOB sdist-make: .../zope.security/setup.py
py27 recreate: .../zope.security/.tox/py27
py27 inst: .../zope.security/.tox/.tmp/package/1/zope.security-5.1.0.dev0.zip
ERROR: invocation failed (exit code 1), logfile: .../zope.security/.tox/py27/log/py27-1.log
============================================================== log start ==============================================================
DEPRECATION: Python 2.7 will reach the end of its life on January 1st, 2020. Please upgrade your Python as Python 2.7 won't be maintained after that date. A future version of pip will drop support for Python 2.7. More details about Python 2 support in pip, can be found at https://pip.pypa.io/en/latest/development/release-process/#python-2-support
Looking in indexes: https://pypi.org/simple, https://mihowitz:****@development.verdi.de/devpi/verdi/prod/+simple/
Processing ./.tox/.tmp/package/1/zope.security-5.1.0.dev0.zip
Requirement already satisfied: setuptools in ./.tox/py27/lib/python2.7/site-packages (from zope.security==5.1.0.dev0) (41.0.1)
Collecting zope.component (from zope.security==5.1.0.dev0)
Collecting zope.i18nmessageid (from zope.security==5.1.0.dev0)
  Using cached https://files.pythonhosted.org/packages/f8/64/b26a46ea05a9167fcd605ec4d2ab7862048bef521b0ddda881d44f775723/zope.i18nmessageid-5.0.0-cp27-cp27m-macosx_10_6_intel.whl
Collecting zope.interface (from zope.security==5.1.0.dev0)
  Using cached https://files.pythonhosted.org/packages/62/57/8268358788e59fc7265cde9c33b7d61d5c072b27c9e80b29cec7f36e5a2d/zope.interface-4.7.1-cp27-cp27m-macosx_10_6_intel.whl
Collecting zope.location (from zope.security==5.1.0.dev0)
  Using cached https://files.pythonhosted.org/packages/9f/3b/4e7a859ebb06911ad118a77528e3c1af018ee5585c832e16151432f7e19b/zope.location-4.2-py2.py3-none-any.whl
Collecting zope.proxy>=4.3.0 (from zope.security==5.1.0.dev0)
  Using cached https://files.pythonhosted.org/packages/a0/13/b7d219f87a76367387680182307871839ef087f345775aeab7780ab34dcd/zope.proxy-4.3.3-cp27-cp27m-macosx_10_6_intel.whl
Collecting zope.schema>=4.2.0 (from zope.security==5.1.0.dev0)
  Using cached https://files.pythonhosted.org/packages/3c/e6/5454a9d72372b73aec715bbded44a0311807dcf7869e2b7bcf196a8e92de/zope.schema-4.9.3-py2.py3-none-any.whl
Collecting Sphinx (from zope.security==5.1.0.dev0)
  Downloading https://development.verdi.de/devpi/root/pypi/%2Bf/f92/9b72e0cfe45fa/Sphinx-2.4.1.tar.gz (5.9MB)
    ERROR: Command errored out with exit status 1:
     command: .../zope.security/.tox/py27/bin/python -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/private/var/folders/_2/5m3t4zvn29gb2f2_4phtk19r0000gn/T/pip-install-4MWhIp/Sphinx/setup.py'"'"'; __file__='"'"'/private/var/folders/_2/5m3t4zvn29gb2f2_4phtk19r0000gn/T/pip-install-4MWhIp/Sphinx/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base pip-egg-info
         cwd: /private/var/folders/_2/5m3t4zvn29gb2f2_4phtk19r0000gn/T/pip-install-4MWhIp/Sphinx/
    Complete output (10 lines):
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/private/var/folders/_2/5m3t4zvn29gb2f2_4phtk19r0000gn/T/pip-install-4MWhIp/Sphinx/setup.py", line 8, in <module>
        import sphinx
      File "sphinx/__init__.py", line 20, in <module>
        from .deprecation import RemovedInNextVersionWarning
      File "sphinx/deprecation.py", line 29
        def deprecated_alias(modname: str, objects: Dict, warning: Type[Warning]) -> None:
                                    ^
    SyntaxError: invalid syntax
    ----------------------------------------
ERROR: Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.
WARNING: You are using pip version 19.2.1, however version 20.0.2 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.

=============================================================== log end ===============================================================
_______________________________________________________________ summary _______________________________________________________________
ERROR:   py27: InvocationError for command .../zope.security/.tox/py27/bin/python -m pip install --exists-action w '.../zope.security/.tox/.tmp/package/1/zope.security-5.1.0.dev0.zip[test,docs]' (exited with code 1)
```

What am I doing wrong?

**Comment by jamadden** [comment]
> What am I doing wrong?

```
Looking in indexes: https://pypi.org/simple, https://mihowitz:****@development.verdi.de/devpi/verdi/prod/+simple/
…

Collecting Sphinx (from zope.security==5.1.0.dev0)
  Downloading https://development.verdi.de/devpi/root/pypi/%2Bf/f92/9b72e0cfe45fa/Sphinx-2.4.1.tar.gz (5.9MB)
```

I'm guessing your custom index might be the problem. It may not support the correct `python_requires` metadata, so pip is installing the wrong version of Sphinx. That should be present in a `data-requires-python` attribute in the index listing.

```console
$ curl https://pypi.org/simple/sphinx/  | grep requires
    <a href="https://files.pythonhosted.org/packages/1b/9f/98d67201c5a6e1aececed03a44a819d0e32adba81414081e303cfaf8c54c/Sphinx-1.7.5-py2.py3-none-any.whl#sha256=85f7e32c8ef07f4ba5aeca728e0f7717bef0789fba8458b8d9c5c294cad134f3" data-requires-python="&gt;=2.7, !=3.0.*, !=3.1.*, !=3.2.*, !=3.3.*">Sphinx-1.7.5-py2.py3-none-any.whl</a><br/>
…
    <a href="https://files.pythonhosted.org/packages/a9/cc/31843aad5b6a1b5a2cdf8ca3332eba0c6ba564bed53981c9a46bb9216bcf/Sphinx-2.0.0b1-py2.py3-none-any.whl#sha256=835c701420102a0a71ba2ed54a5bada2da6fd01263bf6dc8c5c80c798e27709c" data-requires-python="&gt;=3.5">Sphinx-2.0.0b1-py2.py3-none-any.whl</a><br/>
...
```

**Comment by icemac** [comment]
@jamadden You are right, the used devpi server is too old and it seems to be used before https://pypi.org/simple.

Thank you for debugging my local configuration.

**Pull request merged**